docs/RFCS/20200810_leaseholder_cache_invalidation.md

```diff
@@ -0,0 +1,464 @@
+- Feature Name: Range leaseholder cache invalidation
+- Status: draft
+- Start Date: 2020-08-10
+- Authors: knz, nathan, andrei
+- RFC PR: [#52593](https://github.com/cockroachdb/cockroach/pull/52593)
+- Cockroach Issue: [This github card](https://github.com/cockroachdb/cockroach/projects/19#card-36490737),
+  [#50199](https://github.com/cockroachdb/cockroach/issues/50199)
+
+# Summary
+
+This RFC propose an improvement to the range cache to reduce the
+unavailability and performance irregularities that occur when nodes
+are restarted, migrated or simply become unavailable.
+
+To understand the proposal, consider the following scenario:
+
+- consider 3 nodes n1, n2, n3;
+- n1 has a lease initially;
+- n3 knows about n1's lease via its cache;
+- then n1's lease is transfered to n2.
+
+In that scenario, *today* node n3 does not get informed proactively
+that the lease has been transferred to n2. Instead, upon the first
+next request for the range, n3 will first attempt to send the request
+to n1. Then it expects n1 to *redirect* the request to n2. If n1
+happens to still be running, this works; however, if n1 is down (either
+planned or unplanned), the request fails to reach n1, then
+n3 starts trying every other node and discover the lease on n2.
+
+This stall and subsequent discovery process have been observed to
+cause measurable latency blips and other unexpected behavior, which
+this RFC offers to adddress.
+
+To address this, multiple proposals have been considered (detailed
+below).
+
+The selected proposal uses the existing liveness records, and tweaks
+the graceful shutdown process ("drain") to give some time for the
+liveness record to propagate to other nodes.
+
+# Motivation
+
+## Scenarios
+
+This RFC looks at two scenarios:
```

Comment:
For me an important thing is also looking at how quickly caches react to regular cooperative transfers, not just transfers done while draining or new leases issued after a node dies. It'd be nice to eliminate as much bad routing as we can after such a transfer, to help with tail latencies. Like, ideally, no request would ever be routed according to a stale lease. In the particular case of draining, while a node is draining, a lot of requests could be mis-routed according to most of the proposals until the node finally shuts down.